File: 3958-minimum-removals-to-balance-array/3958-minimum-removals-to-balance-array.cpp
class Solution {
public:
    int minRemoval(vector<int>& nums, int k) {
        int n = nums.size();

        if (n == 1)
            return 0;

        sort(nums.begin(), nums.end());

        int i = 0;
        int maxLen = 0;

        for (int j = 1; j < n; j++) {
            while ((long long)nums[j] > (long long)nums[i] * k)
                i++;

            maxLen = max(maxLen, j - i + 1);
        }

        return n - maxLen;
    }
};

File: 3958-minimum-removals-to-balance-array/3958-minimum-removals-to-balance-array.rs
impl Solution {
    pub fn min_removal(mut nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len() as usize;
        nums.sort_unstable();

        let (mut i, mut max_len) = (0, 0);

        let k_long =  k as i64;

        for j in 0..n {
            while (nums[j] as i64) > (nums[i] as i64) * k_long {
                i += 1;
            }

            max_len = max_len.max(j - i + 1);
        }

        (n as i32) - (max_len as i32)
    }
}